ql/experimental/termstructures/yield/multicurvesensitivities.hpp

```diff
@@ -0,0 +1,145 @@
+  to the <em>par quotes</em>, provided in the piecewiseyieldcurve for stripping. If constructed with more
+than one curve, the class iterates over all quotes of the provided curves and shifts each quote of all provided curves taking interdependence into account.
+
+The class computes the sensitvities as a QuantLib Matrix class in the form:
+\f[
+\frac{\partial z_i}{\partial q_j}
+\f]
+where the \f$(z_i)_{i\in \{1,...,n\}}\f$'s are the implied <em>values</em> (being the traits used during curve constructions, e.g. ZeroYield, Discountfactors or ForwardRates) and the the \f$(q_i)_{i\in \{1,...,n\}}\f$'s are the quoted par rates.
+
+
+\note It's the users job to provide all curves that <em>influence</em> the implied rates.
+
+    \ingroup yieldtermstructures
+
+*/
+class MultiCurveSensitivities : public LazyObject {
+private:
+  typedef std::map< std::string, Handle< YieldTermStructure > > curvespec;
+
+public:
+  //! Multi curve sensitivties
+  /*! @param curves std::map of string (curve name) and handle to piecewiseyieldcurve
+  */
+
+  MultiCurveSensitivities(curvespec curves) : curves_(curves) {
+    for (curvespec::const_iterator it = curves_.begin(); it != curves_.end(); ++it)
+      registerWith((*it).second);
+    for (curvespec::const_iterator it = curves_.begin(); it != curves_.end(); ++it) {
+      boost::shared_ptr< PiecewiseYieldCurve< ZeroYield, Linear > > curve =
+          boost::dynamic_pointer_cast< PiecewiseYieldCurve< ZeroYield, Linear > >(it->second.currentLink());
+      QL_REQUIRE(curve != NULL, "Couldn't cast curvename: " << it->first);
+      for (std::vector< boost::shared_ptr< BootstrapHelper< YieldTermStructure > > >::iterator it =
+               curve->instruments_.begin();
+           it != curve->instruments_.end(); ++it) {
+        allQuotes_.push_back((*it)->quote());
+      }
+    }
+  }
+
+  Matrix sensitivities() const;
+  Matrix inverseSensitivities() const;
+
+private:
+  //! \name LazyObject interface
+  //@{
+  void performCalculations() const;
+  //@}
+  // methods
+  std::vector< Real > allZeros() const;
+  std::vector< std::pair< Date, Real > > allNodes() const;
+  mutable std::vector< Rate > origZeros_;
+  std::vector< Handle< Quote > > allQuotes_;
+  std::vector< std::pair< Date, Real > > origNodes_;
+  mutable Matrix sensi_, invSensi_;
+  curvespec curves_;
+};
+
+void MultiCurveSensitivities::performCalculations() const {
+  std::vector< Rate > sensiVector;
+  origZeros_ = allZeros();
+  for (std::vector< Handle< Quote > >::const_iterator it = allQuotes_.begin(); it != allQuotes_.end(); ++it) {
+    Rate bps = 1e-4;
+    Rate origQuote = (*it)->value();
+    boost::shared_ptr< SimpleQuote > q = boost::dynamic_pointer_cast< SimpleQuote >((*it).currentLink());
+    q->setValue(origQuote + bps);
+    std::vector< Rate > tmp(allZeros());
+    for (Size i = 0; i < tmp.size(); ++i)
+      sensiVector.push_back((tmp[i] - origZeros_[i]) / bps);
+    q->setValue(origQuote);
+  }
```

Comment:
I'd write the above as

    q->setValue(origQuote + bps);
    try {
        std::vector< Rate > tmp(allZeros());
        for (Size i = 0; i < tmp.size(); ++i)
            sensiVector.push_back((tmp[i] - origZeros_[i]) / bps);
        q->setValue(origQuote);
    } catch (...) {
        q->setValue(origQuote);
        raise;
    }

so that a failed bootstrap or any other error doesn't leave the curve in a modified state.